### PR TITLE
Fix compiler attributes and functions for IAR and unsupported targets

### PR DESCRIPTION
**Pull request by AlessandroA**
This PR is a modification of [this previously filed PR](https://github.com/ARMmbed/uvisor/pull/210) by @c1728p9 .

It introduces the following changes:
* `ADDRESS_WRITE` is now a static inline function.
* Compiler attributes have been split into core-only and shared.
    * `UVISOR_FORCEINLINE` is now conditionally defined based on the toolchain.
* `vmpu_bits` is only defined with `UVISOR_PRESENT=1`.

@meriac @Patater @niklas-arm @c1728p9 

@c1728p9 Please it would help if you could test this on IAR as we don't have access to it yet.

### CODE REVIEW

File: api/inc/secure_access.h [hunk 1]
```diff
@@ -21,69 +21,70 @@
...
+        case 1:
+            uvisor_write8((volatile uint8_t * volatile) addr, (uint8_t) val);
+            break;
+        default:
+            uvisor_error(USER_NOT_ALLOWED);
+            break;
+    }
+}
+
+#define ADDRESS_WRITE(type, addr, val) __address_write((int) sizeof(type), (volatile uint32_t *) addr, (uint32_t) val)
```
Comment (Patater) [inline api/inc/secure_access.h]: If `sizeof_type` is `size_t`, you can get rid of the `int` cast.

**Comment by Patater** [comment]
LGTM. Nothing I commented about would be a blocker.

**Comment by AlessandroA** [comment]
* Removed cast to `int`.
* Added volatile keyword where missing.

**Comment by Patater** [comment]
LGTM

File: api/inc/secure_access.h [hunk 2]
```diff
@@ -86,4 +52,40 @@ static UVISOR_FORCEINLINE uint8_t uvisor_read8(uint8_t volatile * volatile addr)
     return UVISOR_ASM_MEMORY_ACCESS(ldrb, uint8_t, addr);
 }
 
+/* The switch statement will be optimised away since the compiler already knows
+ * the sizeof_type. */
+static UVISOR_FORCEINLINE void __address_write(size_t sizeof_type, volatile uint32_t *addr, uint32_t val)
```
Comment (meriac) [inline api/inc/secure_access.h]: @AlessandroA : are you confident that a void return type works with IAR ?

File: api/inc/secure_access.h [hunk 3]
```diff
@@ -86,4 +52,40 @@ static UVISOR_FORCEINLINE uint8_t uvisor_read8(uint8_t volatile * volatile addr)
...
+            break;
+    }
+}
+
+#define ADDRESS_WRITE(type, addr, val) __address_write(sizeof(type), (volatile uint32_t *) addr, (uint32_t) val)
+
+/* the conditional statement will be optimised away since the compiler already
+ * knows the sizeof(type) */
+#define ADDRESS_READ(type, addr) \
+    (sizeof(type) == 4 ? uvisor_read32((uint32_t volatile * volatile) (addr)) : \
```
Comment (meriac) [inline api/inc/secure_access.h]: @AlessandroA : turn that into an inline with switch statement, too for consistency ?

File: api/inc/unsupported.h [hunk 4]
```diff
@@ -133,63 +134,64 @@ static inline UVISOR_FORCEINLINE void uvisor_write(uint32_t addr, uint32_t val,
...
+     sizeof(type) == 2 ? uvisor_read16((volatile uint16_t * volatile) (addr)) : \
+     sizeof(type) == 1 ? uvisor_read8((volatile uint8_t * volatile) (addr)) : 0)
+
+/* The switch statement will be optimised away since the compiler already knows
+ * the sizeof_type. */
+static UVISOR_FORCEINLINE void __address_write(size_t sizeof_type, volatile uint32_t *addr, uint32_t val)
+{
+    switch(sizeof_type) {
+        case 4:
+            uvisor_write32((volatile uint32_t * volatile) addr, (uint32_t) val);
```
Comment (c1728p9) [inline api/inc/unsupported.h]: Why are both the pointer and the value being pointed to volatile?  Only the value being pointed to is in volatile memory.  This also introduces the following warning on IAR:
[Warning] unsupported.h@180: [Pe191]: type qualifier is meaningless on cast type

File: api/inc/uvisor_exports.h [hunk 5]
```diff
@@ -34,15 +34,15 @@
...
+#else
+#define UVISOR_FORCEINLINE inline __attribute__((always_inline))
+#endif
 #define UVISOR_PACKED      __attribute__((packed))
 #define UVISOR_WEAK        __attribute__((weak))
-#define UVISOR_ALIAS(f)    __attribute__((weak, alias (#f)))
-#define UVISOR_LINKTO(f)   __attribute__((alias (#f)))
 #define UVISOR_NORETURN    __attribute__((noreturn))
-#define UVISOR_NAKED       __attribute__((naked))
 #define UVISOR_RAMFUNC     __attribute__ ((section (".ramfunc"), noinline))
```
Comment (c1728p9) [inline api/inc/uvisor_exports.h]: These attributes are invalid for IAR.  You'll need to define them inside __ICCARM__ as well

**Comment by c1728p9** [comment]
I gave the latest uvisor-mbed-lib a try, and I'm still getting errors.  I think this patch missed a bit from the one it was based on.

File: api/inc/unsupported.h (hunk 4 above)
Comment (AlessandroA) [inline api/inc/unsupported.h]: This is due to an issue, reported on GCC. Without the `volatile` compiler optimizations for the read/write functions were preventing our inline assembly code to be compiled as desired. It was fixed in #172 and has #173 has been opened to keep track of it.

But now that I think about it, the change here was only introduced for consistency with the supported case — it's not really needed and it shouldn't be there. I will remove this in a new PR.

**Pull request merged**